### DISCUSSION TITLE
OpenEdge VSCode configuration file

### DISCUSSION THREAD

**Discussion by gquerret** [vscode-abl/vscode-abl#20] [Ideas]
As it's still possible to break the configuration file format without impacting too many people, I'd like to have some input from the community about names and entries in the VSCode configuration file for OE projects. Feel free to comment anything, nothing is written in stone (yet).

**File name**: `openedge-project.json`
I don't want name to start with a dot, as the file has to be part of the repository and has to be easily accessible, so no reason to hide it. Json format, XML is too verbose, and YAML makes my eyes bleed.

## Attributes

* **name** + **version**: Name and version number of the project. Will be used later for dependency management. Can't be undefined.
* **oeversion**: OpenEdge version number. Should reference a version number configured in VSCode.
* **graphicalMode**: use prowin or _progres. Name is coming from PCT, maybe time to switch to something better...
*  **numThreads**: number of OpenEdge background sessions. Currently has to be a number, no variable expansion. That could be implemented later.
* **extraParameters**: always passed as parameters to the OpenEdge sessions
* **initProcedures**: array of procedures to be instantiated at OE session startup
* **dbConnections**: array of objects with those properties:
  * *name*: connection name
  * *dumpFile*: comma-separated list of DF files for this database
  * *connect*: ABL connection string. Logical name of the database should match the connection name
  * *aliases*: optional array of strings, one entry per alias name
* **buildDirectory**: global build directory (where rcode is created) in case it's not specified per source directory. Array list.
* **buildPath**: array of object with those properties:
  * *type*: can be `source` or `propath`. Could be `project` in the future.
  * *path*: relative-path only for source directories (and has to be a subdirectory of the project directory). Can be an absolute path for propath entries
  * *buildDir*: only for source directories. Defaults to source directory if global build directory is not defined.
  * *xrefDir*: stores XREF file
* **custom**: per Mike's request ( #15 ) , a specific section can be added for customization of the project. This attribute is an object, and its content will never be read by the VSCode plugin.

## Additional profiles

The need for additional profiles is highlighted by these use cases:
* Project has a GUI and a TTY version ; source directories are identical, but include / excludes patterns can be different, and the runtime is different in both cases
* Different database connections when working from home, working with Docker, or accessing a shared set of DB on the network
* Small parts of the code require an additional DB connection, or an additional entry in the propath
* Running project with different OE versions

Additional profiles will be stored in a **profiles** attributes (array-type), each entry being an object with two attributes, **name** and **value**. The *value* attribute is also an object with all standard attributes (*oeversion*, *graphicalMode*, ...).
❓ Would it be useful to have a **inherits** attribute ? Or better to specify all attributes manually ? 
The current profile will be stored in a JSON file in the root directory (`$ROOT_DIR/profile.json`). The profile name will used to identify the profile ; if file is not present (or invalid reference), the default profile is used. IMO, this file should never be pushed in the code repository.
An entry point in VSCode will allow switching from a profile to another one; that will restart the language server, and refresh the VSCode configuration for source entries.

**Comment by stefandrissen** [comment]
I think most companies need to support multiple OpenEdge versions for the same codebase, I'm not sure how this is handled and / or configured by other (vs code) language servers.

**Reply by gquerret** [reply]
The Java version number is specified in the pom.xml (for Maven) or build.gradle (for Gradle). The LS then use the installation directory configured on the machine where VSCode is running.
My opinion is that the version number should be stored in the config file, which is shared in the code repository. Maintaining multiple versions for the same codebase usually means that a separate branch is created. The config file with a different OE version number can then 	also be stored directly in the branch.

**Reply by DustinDunbar** [reply]
I wouldn't say we need to have a different source for each oeversion as in it isn't a different version/branch - we just need to support building in both oe11.7 and 12x at least while it takes a few years to migrate.  Right now for example our db's are oe11 and front end runs on oe12

**Reply by gquerret** [reply]
Build with OE 11.x and 12.x is easily done by using Ant / PCT + Jenkins pipelines (or any other CI tool). You can use extra profiles in the LS if you want to switch your local dev environment from 11 to 12 and vice-versa.

**Comment by gquerret** [comment]
Each build path entry should allow includes / excludes pattern, such as:
```
// Simple way
"buildPath": [
  { "type": "source", path: "src/main/abl" includes="**/*.p,**/*.w" excludes="**/*.f" }
]
```

Or 
```
// More complex, but is that really required ?
"buildPath": [
  { "type": "source", path: "src/main/abl", "patterns": [
      { "type": "include", "pattern": "**/*.p,**/*.w"},  { "type": "exclude", "pattern": "**/*.f"}
    ]}
]
```

**Comment by PeterJudgeZA** [comment]
Is a buildPath entry of type "propath" really just a procedure library?

**Reply by gquerret** [reply]
No, propath entries can be directories or PL files.

**Reply by gquerret** [reply]
Source entries are monitored, and any file change will trigger a recompile of the file. Propath entries are also monitored, but can only trigger dependency recompile.

**Comment by PeterJudgeZA** [comment]
It seems like the `oeversion` property should point to a runtime configuration (and so should be named that way). The runtime configuration  should declare
- $DLC, exe (prowin/_progress
- cpinternal and cpstream
- graphicalMode
- `-T` /temp-dir 
- other command-line-only options

These configs can also then be used by test launch cofnigs.

The project config should reference one of these for compilation/build purposes, and allow overrides for things like
- propath
- db connections
- etc

**Comment by gquerret** [comment]
Workspace configuration is currently just the list of OE version number, with the install directory, and each project has to reference the version number.
* Cpstream: I think this should be read from the VSCode config, as it's required for proper display. VSCode config is in .vscode/settings.json (IIRC)
* CpInternal: the background sessions will use utf-8. 
* TempDir: background sessions will use a specific directory (controlled by language server)

The background sessions are only for the language server. Would you specify more info at the workspace level, and reference that in the project config file ?

Launch configurations should be controlled a bit differently, where propath and db connections can be overridden. Config should be stored in the standard .vscode/launch.json file. Should inherit the default DB and propath coming from the project, with a possibility of overriding those properties.

**Comment by gquerret** [comment]
Just added section about extra profiles in the configuration file.

**Reply by gquerret** [reply]
An example configuration file with a default OE12 profile, and an additional OE11 profile: https://github.com/Riverside-Software/pct/blob/master/openedge-project.json
Different build directory, different set of excluded files, different DB connection, all of that while still in the same coding environment.

Language server reads profile.json in root directory (if available) ; VSCode command will be available to switch a project to another profile.
```
{"profile": "v11"}
```

**Comment by sam-vline** [comment]
It would be great if there was a way to add PREPROCESS to the compile statement and specify an output file. I'm working on a project that uses a framework with a lot of preprocessor commands and trying to figure out/second guess what code is getting generated is making my head swim

**Reply by gquerret** [reply]
You can use the "Preprocess current file" action on the current buffer. And please open new discussions / tickets rather than using existing ones.
